### PR TITLE
Add support to invalidate cookies during authentication flow failure

### PR DESCRIPTION
**Pull request by mevan-karu**
### Proposed changes in this pull request
- This PR adds support to invalidate cookies that are used in authentication flow, when an user is redirected to `retry.do` due to an server or client error.
- Partially fixes : https://github.com/wso2-enterprise/asgardeo-product/issues/1643

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityConfigParser.java [hunk 1]
```diff
@@ -204,6 +212,25 @@ private void buildConfiguration() {
...
+        OMElement cookiesToInvalidate = this.getConfigElement(IdentityConstants.COOKIES_TO_INVALIDATE_CONFIG);
+        Iterator<OMElement> cookies = cookiesToInvalidate.getChildrenWithName(
+                new QName(IdentityCoreConstants.IDENTITY_DEFAULT_NAMESPACE, IdentityConstants.COOKIE));
+        if (cookies != null) {
+            while (cookies.hasNext()) {
+                OMElement cookie = cookies.next();
+                String cookieName = cookie.getAttributeValue(new QName(IdentityConstants.COOKIE_NAME));
+
+                if (StringUtils.isBlank(cookieName)) {
+                    throw IdentityRuntimeException.error("Cookie name not defined correctly");
```
Comment (IsuraD) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityConfigParser.java:226]: Invalid Configuration.  Cookie name is not defined correctly"

**Review by IsuraD — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 2]
```diff
@@ -531,31 +532,78 @@ private static CallBackHandlerFactory getCallBackHandlerFactory() {
...
+                uriBuilder.addParameter("statusMsg", statusMsg);
+            }
             request.setAttribute(FrameworkConstants.RequestParams.FLOW_STATUS, AuthenticatorFlowStatus.INCOMPLETE);
-            response.sendRedirect(getRedirectURL(uriBuilder.build().toString(), request));
+            if (context != null) {
+                if (IdentityTenantUtil.isTenantedSessionsEnabled()) {
+                    uriBuilder.addParameter(USER_TENANT_DOMAIN_HINT, context.getUserTenantDomain());
+                }
+                uriBuilder.addParameter(REQUEST_PARAM_SP, context.getServiceProviderName());
+                uriBuilder.addParameter(TENANT_DOMAIN, context.getTenantDomain());
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:588]: let's add tenant domain only if ((!IdentityTenantUtil.isTenantQualifiedUrlsEnabled())
https://github.com/wso2/carbon-identity-framework/pull/3373/files#diff-c827402ed6c15cc879db866bf51cbd2c886fd806013a4c5ad06c9b47bb3a8298L584

Comment (mevan-karu) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java]: Fixed in https://github.com/wso2/carbon-identity-framework/pull/3373/commits/2f9dc5c391cb14eb4538bda7de2b38da9c3c17c5

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityConfigParser.java (hunk 1 above)
Comment (mevan-karu) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityConfigParser.java]: Fixed in https://github.com/wso2/carbon-identity-framework/pull/3373/commits/2f9dc5c391cb14eb4538bda7de2b38da9c3c17c5

**Review by senthalan — APPROVED** [review]

**Pull request merged**